text/0000-show-string-guidelines.md

```diff
@@ -0,0 +1,166 @@
+
+The `fmt::Display` trait is intended for user-facing output. It should:
+
+* Be implemented for scalars, strings, and other basic types.
+* Be implemented for generic wrappers like `Option<T>` or smart pointers, where
+  the output can be wholly delegated to a *single* `fmt::Display` implementation
+  on the underlying type.
+* *Not* be implemented for generic containers like `Vec<T>` or even `Result<T, E>`,
+  where there is no useful, general way to tailor the output for user consumption.
+* Be implemented for *specific* user-defined types as useful for an application,
+  with application-defined user-facing output. In particular, applications will
+  often make their types implement `fmt::Display` specifically for use in
+  `format` interpolation.
+* Never panic.
+* Avoid quotes, escapes, and so on unless specifically desired for a user-facing purpose.
+* Require use of an explicit adapter (like the `display` method in
+  `Path`) when it potentially looses significant information.
+
+A common pattern for `fmt::Display` is to provide simple "adapters", which are
+types wrapping another type for the sole purpose of formatting in a certain
+style or context. For example:
+
+```rust
+pub struct ForHtml<'a, T>(&'a T);
+pub struct ForCli<'a, T>(&'a T);
+
+impl MyInterestingType {
+    fn for_html(&self) -> ForHtml<MyInterestingType> { ForHtml(self) }
+    fn for_cli(&self) -> ForCli<MyInterestingType> { ForCli(self) }
+}
+
+impl<'a> fmt::Display for ForHtml<'a, MyInterestingType> { ... }
+impl<'a> fmt::Display for ForCli<'a, MyInterestingType> { ... }
+```
+
+## Rationale for format specifiers
+
+Given the above conventions, it should be clear that `fmt::Debug` is
+much more commonly *implemented* on types than `fmt::Display`. Why,
+then, use `{}` for `fmt::Display` and `{:?}` for `fmt::Debug`? Aren't
+those the wrong defaults?
+
+There are two main reasons for this choice:
+
+* Debugging output usually makes very little use of interpolation. In general,
+  one is typically using `#[derive(Show)]` or `format!("{:?}",
+  something_to_debug)`, and the latter is better done via
+  [more direct convenience](https://github.com/SimonSapin/rust-std-candidates#the-show-debugging-macro).
+
+* When creating tailored string output via interpolation, the expected "default"
+  formatting for things like strings is unquoted and unescapted. It would be
+  surprising if the default specifiers below did not yield `"hello, world!" as the
+  output string.
+
+  ```rust
+  format!("{}, {}!", "hello", "world")
+  ```
+
+In other words, although more types implement `fmt::Debug`, most
+meaningful uses of interpolation (other than in such implementations)
+will use `fmt::Display`, making `{}` the right choice.
+
+## Use in errors
+
+Right now, the (unstable) `Error` trait comes equipped with a `description`
+method yielding an `Option<String>`. This RFC proposes to drop this method an
+instead inherit from `fmt::Display`. It likewise proposes to make `unwrap` in
```

Comment:
Due to the common usage of `Error` as a trait object, this may be difficult to get enough bang for our buck. For example what we'd want to have is something like:

```rust
let e: &Error = ...;
println!("{}", e);
```

Today, however, this generates an ICE, but I think if the ICE were fixed that it would require an impl to look like:

```rust
impl Display for Error {
   // ???
}
```

Right now there is a `Display` implementation for `&Display`, but we don't allow coercions between trait objects (can't cast `&Error` to `&Display`). Until we allow that, I think we'd have to do the following for now:

```rust
trait Error: Display {
    fn to_string(&self) -> String {
        format!("{}", self)
    }
}
```

(but I think this is just a minor technical detail)